File: backend/models/schemas.py
"""
Synapsis Backend — Pydantic Schemas
Data contracts from ARCHITECTURE.md §9.
"""

from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Ingestion
# ---------------------------------------------------------------------------

class IngestionRecord(BaseModel):
    ingestion_id: str
    source_type: str  # "auto_scan" | "watcher_event"
    modality: str  # "text" | "pdf" | "image" | "audio" | "json"
    source_uri: str
    collected_at: str
    checksum: str
    status: str  # "queued" | "processed" | "failed" | "skipped"


# ---------------------------------------------------------------------------
# Knowledge Card
# ---------------------------------------------------------------------------

class KnowledgeCard(BaseModel):
    card_id: str
    title: str
    summary: str
    category: str
    entities: list[str] = Field(default_factory=list)
    action_items: list[str] = Field(default_factory=list)
    modality: str
    event_time: str
    ingestion_time: str
    source_uri: str | None = None


# ---------------------------------------------------------------------------
# Retrieval / Reasoning
# ---------------------------------------------------------------------------

class ChunkEvidence(BaseModel):
    chunk_id: str
    file_name: str
    snippet: str
    score_dense: float = 0.0
    score_sparse: float = 0.0
    score_final: float = 0.0


class AnswerPacket(BaseModel):
    answer: str
    confidence: str  # "high" | "medium" | "low" | "none"
    confidence_score: float
    uncertainty_reason: str | None = None
    sources: list[ChunkEvidence] = Field(default_factory=list)
    verification: str  # "APPROVE" | "REVISE" | "REJECT"
    reasoning_chain: str | None = None


# ---------------------------------------------------------------------------
# Query request
# ---------------------------------------------------------------------------

class QueryRequest(BaseModel):
    question: str
    top_k: int = 10
    include_graph: bool = True


# ---------------------------------------------------------------------------
# Graph visualization
# ---------------------------------------------------------------------------

class GraphNode(BaseModel):
    id: str
    type: str
    name: str
    properties: dict | None = None
    mention_count: int = 1


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    relationship: str
    properties: dict | None = None


class GraphData(BaseModel):
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Timeline
# ---------------------------------------------------------------------------

class TimelineItem(BaseModel):
    id: str
    title: str
    summary: str | None = None
    category: str | None = None
    modality: str
    source_uri: str | None = None
    ingested_at: str
    entities: list[str] = Field(default_factory=list)


class TimelineResponse(BaseModel):
    items: list[TimelineItem] = Field(default_factory=list)
    total: int = 0
    page: int = 1
    page_size: int = 20


# ---------------------------------------------------------------------------
# Config / Sources
# ---------------------------------------------------------------------------

class SourceConfig(BaseModel):
    id: str | None = None
    path: str
    enabled: bool = True
    exclude_patterns: list[str] = Field(default_factory=list)


class SourcesConfigResponse(BaseModel):
    watched_directories: list[SourceConfig] = Field(default_factory=list)
    exclude_patterns: list[str] = Field(default_factory=list)
    max_file_size_mb: int = 50
    scan_interval_seconds: int = 30
    rate_limit_files_per_minute: int = 10


class SourcesConfigUpdate(BaseModel):
    watched_directories: list[str] = Field(default_factory=list)
    exclude_patterns: list[str] = Field(default_factory=list)
    max_file_size_mb: int | None = None
    scan_interval_seconds: int | None = None
    rate_limit_files_per_minute: int | None = None


# ---------------------------------------------------------------------------
# Health
# ---------------------------------------------------------------------------

class ServiceStatus(BaseModel):
    status: str  # "up" | "down"
    detail: dict | None = None


class HealthResponse(BaseModel):
    status: str  # "healthy" | "degraded" | "unhealthy"
    ollama: ServiceStatus
    qdrant: ServiceStatus
    sqlite: ServiceStatus
    disk_free_gb: float | None = None
    uptime_seconds: float = 0.0


# ---------------------------------------------------------------------------
# Ingestion Status
# ---------------------------------------------------------------------------

class IngestionStatusResponse(BaseModel):
    queue_depth: int = 0
    files_processed: int = 0
    files_failed: int = 0
    files_skipped: int = 0
    last_scan_time: str | None = None
    is_watching: bool = False
    watched_directories: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Insights / Digest
# ---------------------------------------------------------------------------

class InsightItem(BaseModel):
    type: str  # "connection" | "contradiction" | "pattern" | "digest"
    title: str
    description: str
    related_entities: list[str] = Field(default_factory=list)
    created_at: str


class DigestResponse(BaseModel):
    insights: list[InsightItem] = Field(default_factory=list)
    generated_at: str | None = None


# ---------------------------------------------------------------------------
# Memory detail
# ---------------------------------------------------------------------------

class MemoryDetail(BaseModel):
    id: str
    filename: str
    modality: str
    source_uri: str | None = None
    ingested_at: str
    status: str
    summary: str | None = None
    category: str | None = None
    entities: list[str] = Field(default_factory=list)
    action_items: list[str] = Field(default_factory=list)
    chunks: list[dict] = Field(default_factory=list)


class MemoryStats(BaseModel):
    total_documents: int = 0
    total_chunks: int = 0
    total_nodes: int = 0
    total_edges: int = 0
    categories: dict[str, int] = Field(default_factory=dict)
    modalities: dict[str, int] = Field(default_factory=dict)
    entity_types: dict[str, int] = Field(default_factory=dict)

File: frontend/synapsis/lib/api/schemas.ts
/**
 * Zod runtime schemas for backend response validation.
 * Mirror of types/contracts.ts, used by API client for parse-safe responses.
 *
 * Source: ARCHITECTURE.md, BACKEND_CONTRACT_ALIGNMENT.md
 */

import { z } from "zod";

// ─── Enums ───

export const ConfidenceLevelSchema = z.enum(["high", "medium", "low", "none"]);
export const VerificationStatusSchema = z.enum(["APPROVE", "REVISE", "REJECT"]);
export const ServiceStateSchema = z.enum(["up", "down"]);
export const HealthStateSchema = z.enum(["healthy", "degraded", "unhealthy"]);
export const TimelineModalitySchema = z.enum(["text", "pdf", "image", "audio", "json"]);
export const IngestionWsEventTypeSchema = z.enum([
  "status",
  "file_processed",
  "file_deleted",
  "file_error",
  "scan_started",
  "scan_completed",
  "incident",
]);

// ─── Evidence ───

export const ChunkEvidenceSchema = z.object({
  chunk_id: z.string(),
  file_name: z.string(),
  snippet: z.string(),
  score_dense: z.number(),
  score_sparse: z.number(),
  score_final: z.number(),
});

// ─── Answer ───

export const AnswerPacketSchema = z.object({
  answer: z.string(),
  confidence: ConfidenceLevelSchema,
  confidence_score: z.number(),
  uncertainty_reason: z.string().nullable(),
  sources: z.array(ChunkEvidenceSchema),
  verification: VerificationStatusSchema,
  reasoning_chain: z.string().nullable(),
});

// ─── Query Stream ───

export const QueryStreamMessageSchema = z.discriminatedUnion("type", [
  z.object({ type: z.literal("token"), data: z.string() }),
  z.object({ type: z.literal("done"), data: AnswerPacketSchema }),
  z.object({ type: z.literal("error"), data: z.string() }),
]);

// ─── Graph ───

export const GraphNodeSchema = z.object({
  id: z.string(),
  type: z.string(),
  name: z.string(),
  properties: z.record(z.unknown()).nullable(),
  mention_count: z.number(),
});

export const GraphEdgeSchema = z.object({
  id: z.string(),
  source: z.string(),
  target: z.string(),
  relationship: z.string(),
  properties: z.record(z.unknown()).nullable(),
});

export const GraphDataSchema = z.object({
  nodes: z.array(GraphNodeSchema),
  edges: z.array(GraphEdgeSchema),
});

// ─── Timeline ───

export const TimelineItemSchema = z.object({
  id: z.string(),
  title: z.string(),
  summary: z.string().nullable(),
  category: z.string().nullable(),
  modality: TimelineModalitySchema,
  source_uri: z.string().nullable(),
  ingested_at: z.string(),
  entities: z.array(z.string()),
});

export const TimelineResponseSchema = z.object({
  items: z.array(TimelineItemSchema),
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
});

// ─── Memory Detail ───

export const MemoryDetailChunkSchema = z.object({
  id: z.string(),
  content: z.string(),
  chunk_index: z.number(),
});

export const MemoryDetailSchema = z.object({
  id: z.string(),
  filename: z.string(),
  modality: z.string(),
  source_uri: z.string().nullable(),
  ingested_at: z.string(),
  status: z.string(),
  enrichment_status: z.string().nullable().optional(),
  summary: z.string().nullable(),
  category: z.string().nullable(),
  entities: z.array(z.string()),
  action_items: z.array(z.string()),
  chunks: z.array(z.record(z.unknown())),
});

// ─── Memory Stats ───

export const MemoryStatsSchema = z.object({
  total_documents: z.number(),
  total_chunks: z.number(),
  total_nodes: z.number(),
  total_edges: z.number(),
  categories: z.record(z.number()),
  modalities: z.record(z.number()),
  entity_types: z.record(z.number()),
});

export const MemorySearchResultSchema = z.object({
  chunk_id: z.string(),
  content: z.string(),
  chunk_index: z.number(),
  summary: z.string().nullable(),
  category: z.string().nullable(),
  action_items: z.array(z.string()),
  document_id: z.string(),
  filename: z.string(),
  modality: TimelineModalitySchema,
  ingested_at: z.string(),
});

// ─── Config ───

export const WatchedDirectorySchema = z.object({
  id: z.string().nullable(),
  path: z.string(),
  enabled: z.boolean(),
  exclude_patterns: z.array(z.string()),
});

export const SourcesConfigResponseSchema = z.object({
  watched_directories: z.array(WatchedDirectorySchema),
  exclude_patterns: z.array(z.string()),
  max_file_size_mb: z.number(),
  scan_interval_seconds: z.number(),
  rate_limit_files_per_minute: z.number(),
});

// ─── Ingestion ───

export const IngestionStatusResponseSchema = z.object({
  queue_depth: z.number(),
  files_processed: z.number(),
  files_failed: z.number(),
  files_skipped: z.number(),
  last_scan_time: z.string().nullable(),
  is_watching: z.boolean(),
  watched_directories: z.array(z.string()),
});

export const IngestionScanResponseSchema = z.object({
  message: z.string(),
  files_processed: z.number(),
  errors: z.number(),
});

export const IngestionWsMessageSchema = z.object({
  event: IngestionWsEventTypeSchema,
  payload: z.record(z.unknown()),
});

export const IncidentSeveritySchema = z.enum(["info", "warning", "error", "critical"]);

export const RuntimeIncidentSchema = z.object({
  id: z.string(),
  timestamp: z.string(),
  subsystem: z.string(),
  operation: z.string(),
  reason: z.string(),
  severity: IncidentSeveritySchema,
  blocked: z.boolean(),
  payload: z.record(z.unknown()).nullable(),
});

export const RuntimePolicyResponseSchema = z.object({
  fail_fast: z.boolean(),
  allow_model_fallback: z.boolean(),
  lane_assignment: z.record(z.string()),
  outage_policy: z.string(),
});

// ─── Health ───

export const ServiceHealthDetailSchema = z.object({
  status: ServiceStateSchema,
  detail: z.record(z.unknown()).nullable(),
});

export const HealthResponseSchema = z.object({
  status: HealthStateSchema,
  ollama: ServiceHealthDetailSchema,
  qdrant: ServiceHealthDetailSchema,
  sqlite: ServiceHealthDetailSchema,
  disk_free_gb: z.number().nullable(),
  uptime_seconds: z.number(),
});

// ─── Insights ───

export const InsightItemSchema = z.object({
  type: z.string(),
  title: z.string().optional(),
  description: z.string(),
  related_entities: z.array(z.string()).optional(),
  created_at: z.string().optional(),
  entities: z.array(z.string()).optional(),
});

export const DigestResponseSchema = z.object({
  insights: z.array(InsightItemSchema),
  generated_at: z.string().nullable(),
});

export const PatternsResponseSchema = z.object({
  patterns: z.array(InsightItemSchema),
});
